spec/validators/budget_date_validator_spec.rb

```diff
@@ -0,0 +1,66 @@
+require "rails_helper"
+
+RSpec.describe BudgetDatesValidator do
+  subject { build(:budget) }
+
+  context "when the start date is the same as the end date" do
+    it "is valid" do
+      subject.period_start_date = Date.today
+      subject.period_end_date = Date.today
+
+      expect(subject).to be_valid
+    end
+  end
+
+  context "when the start date is before the end date" do
+    it "is valid" do
+      subject.period_start_date = Date.yesterday
+      subject.period_end_date = Date.today
+
+      expect(subject).to be_valid
+    end
+  end
+
+  context "when the start date is after the end date" do
+    it "is not valid" do
+      subject.period_start_date = Date.tomorrow
+      subject.period_end_date = Date.today
+
+      expect(subject).not_to be_valid
+    end
+
+    it "adds an error message to the :period_start_date" do
+      subject.period_start_date = Date.tomorrow
+      subject.period_end_date = Date.today
+
+      subject.valid?
+      expect(subject.errors.messages[:period_start_date]).to include I18n.t("activerecord.errors.models.budget.attributes.period_start_date.not_after_end_date")
+    end
+  end
+
+  context "when the dates are within one year of each other" do
+    it "is valid" do
+      subject.period_start_date = Date.today
+      subject.period_end_date = subject.period_start_date + 6.months
```

Comment:
We've got really good test coverage for the cases when:

- dates are both the same
- dates are 2 day apart
- dates are 6 months apart

To complete the set could we have a final test that check what happens if the period is exactly 365 days apart? Then I wonder about how we deal with leap years and if that's a problem too!